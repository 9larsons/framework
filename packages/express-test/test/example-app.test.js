const {assert} = require('./utils');

const Agent = require('../'); // we require the root file
const app = require('../example/app');
const {before} = require('mocha');

let agent;

/**
 * These functions are async because in a real world scenario,
 * the function that loads "app" would be async.
 * This has a huge impact as await will call .then if available
 * this is the entire reason for splitting Agent and Request
 * So we want to test it is working!
 */

async function getAgent() {
    return new Agent(app);
}

async function getAPIAgent() {
    return new Agent(app, {baseUrl: '/api/'});
}

async function getExtendedAPIAgent() {
    agent = await getAPIAgent();

    agent.login = async function () {
        return await agent.post('/session/', {
            body: {
                username: 'hello',
                password: 'world'
            }
        });
    };

    return agent;
}

describe('Example App', function () {
    before(async function () {
        agent = await getAgent();
    });

    describe('Object Destructuring', function () {
        it('Basic test of GET /', async function () {
            try {
                const {statusCode, text} = await agent.get('/');
                assert.equal(statusCode, 200);
                assert.equal(text, 'Hello World!');
            } catch (error) {
                assert.fail(`Should not have thrown an error', but got ${error.stack}`);
            }
        });

        describe('API Agent with authentication in two steps', function () {
            before(async function () {
                agent = await getAPIAgent();
            });

            it('cannot perform request without session', async function () {
                const {statusCode, headers, body, text} = await agent.get('/foo/');

                assert.equal(statusCode, 403);
                assert.deepEqual(Object.keys(headers), ['x-powered-by', 'content-type', 'content-length', 'etag']);
                assert.deepEqual(body, {});
                assert.equal(text, 'Forbidden');
            });

            it('create session & make authenticated request', async function () {
                const sessionRes = await agent.post('/session/', {
                    body: {
                        username: 'hello',
                        password: 'world'
                    }
                });

                assert.equal(sessionRes.statusCode, 200);
                assert.deepEqual(Object.keys(sessionRes.headers), ['x-powered-by', 'content-type', 'content-length', 'etag', 'set-cookie']);
                assert.deepEqual(sessionRes.body, {});
                assert.equal(sessionRes.text, 'OK');

                const {statusCode, headers, body, text} = await agent.get('/foo/');

                assert.equal(statusCode, 200);
                assert.deepEqual(Object.keys(headers), ['x-powered-by', 'content-type', 'content-length', 'etag']);
                assert.deepEqual(body, {foo: [{bar: 'baz'}]});
                assert.equal(text, '{"foo":[{"bar":"baz"}]}');
            });
        });

        describe('API Agent with login function', function () {
            before(async function () {
                agent = await getExtendedAPIAgent();
                await agent.login();
            });

            it('make an authenticated request', async function () {
                const {statusCode, headers, body, text} = await agent.get('/foo/');

                assert.equal(statusCode, 200);
                assert.deepEqual(Object.keys(headers), ['x-powered-by', 'content-type', 'content-length', 'etag']);
                assert.deepEqual(body, {foo: [{bar: 'baz'}]});
                assert.equal(text, '{"foo":[{"bar":"baz"}]}');
            });
        });

        describe('headers and body', function () {
            before(async function () {
                agent = await getAgent();
            });

            it('check headers and body using reqOptions', async function () {
                const {statusCode, headers, body} = await agent
                    .post('/check/', {
                        body: {foo: 'bar'},
                        headers: {'x-check': true}
                    });

                assert.equal(statusCode, 200);
                assert.deepEqual(body, {foo: 'bar'});
                assert.equal(headers['x-checked'], 'true');
            });

            it('check headers and body using set chaining', async function () {
                const {statusCode, headers, body} = await agent
                    .post('/check/')
                    .body({foo: 'bar'})
                    .header('x-check', true);

                assert.equal(statusCode, 200);
                assert.deepEqual(body, {foo: 'bar'});
                assert.equal(headers['x-checked'], 'true');
            });

            it('check headers and body using set and expect chaining', async function () {
                const {body} = await agent
                    .post('/check/')
                    .body({foo: 'bar'})
                    .header('x-check', true)
                    .expectStatus(200)
                    .expectHeader('x-checked', 'true');

                assert.deepEqual(body, {foo: 'bar'});
            });
        });
    });
});
