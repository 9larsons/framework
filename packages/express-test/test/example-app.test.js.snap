// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example App Object Destructuring headers, status and body check body using snapshot matching errors correctly 1 1`] = `
Object {
  "foo": "bar",
  "id": "0.5su8igllk79",
}
`;

exports[`Example App Object Destructuring headers, status and body check body using snapshot matching errors correctly for random data 1 1`] = `
Object {
  "foo": "bar",
  "id": "0.kr2ostne6qd",
}
`;

exports[`Example App Object Destructuring headers, status and body check body using snapshot matching properties works for random data 1 1`] = `
Object {
  "foo": "bar",
  "id": Any<String>,
}
`;

exports[`Example App Object Destructuring headers, status and body check headers, status and body using set, expect chaining & snapshot matching 1 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`Example App Object Destructuring headers, status and body check headers, status and body using set, expect chaining & snapshot matching 2 1`] = `
Object {
  "content-length": "13",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"d-pedE0BZFQNM7HX6mFsKPL6l+dUo\\"",
  "x-checked": "true",
  "x-powered-by": "Express",
}
`;
